packages/components/src/QueryModel/QueryModel.ts

```diff
@@ -0,0 +1,307 @@
+import { List } from 'immutable';
+import { immerable } from 'immer';
+import { Filter, Query } from '@labkey/api';
+
+import { naturalSort, QueryColumn, QueryInfo, SchemaQuery, ViewInfo } from '..';
+import { QuerySort } from '../components/base/models/model';
+import { getOrDefault } from './utils';
+
+export enum LoadingState {
+    // The model has been initialized but not loaded
+    INITIALIZED = 'INITIALIZED',
+    // The model is currently loading
+    LOADING = 'LOADING',
+    // The model is loaded
+    LOADED = 'LOADED',
+}
+
+/**
+ * Creates a QueryModel ID for a given SchemaQuery. The id is just the SchemaQuery snake-cased as
+ * schemaName-queryName-viewName or schemaName-queryName if viewName is undefined.
+ *
+ * @param schemaQuery: SchemaQuery
+ */
+export function createQueryModelId(schemaQuery: SchemaQuery): string {
+    const { schemaName, queryName, viewName } = schemaQuery;
+    return `${schemaName}-${queryName}${viewName !== undefined ? '-' + viewName : ''}`;
+}
+
+const sortStringMapper = (s: QuerySort): string => s.toRequestString();
+
+export interface GridMessage {
+    area: string;
+    type: string;
+    content: string;
+}
+
+export interface QueryConfig {
+    baseFilters?: Filter.IFilter[];
+    containerFilter?: Query.ContainerFilter;
+    containerPath?: string;
+    id?: string;
+    includeDetailsColumn?: boolean;
+    includeUpdateColumn?: boolean;
+    keyValue?: any; // TODO: better name
+    maxRows?: number;
+    offset?: number;
+    omittedColumns?: string[];
+    queryParameters?: { [key: string]: any};
+    requiredColumns?: string[];
+    schemaQuery: SchemaQuery;
+    sorts?: QuerySort[];
+}
+
+export interface IQueryModel extends QueryConfig {
+    error?: string;
+    // Separate from baseFilters because these are set by the user when interacting with grids (e.g. via omnibox)
+    filterArray: Filter.IFilter[];
+    // Set by server (Assay QC, etc)
+    messages: GridMessage[];
+    queryInfo?: QueryInfo;
+    queryInfoLoadingState: LoadingState;
+    orderedRows?: string[];
+    rows?: { [key: string]: any};
+    rowCount?: number;
+    rowsLoadingState: LoadingState;
+}
+
+const DEFAULT_OFFSET = 0;
+const DEFAULT_MAX_ROWS = 20;
+
+export class QueryModel implements IQueryModel {
```

Comment:
I think `immer` can be leveraged further to allow for `QueryModel` to be a truly immutable type. I'm using this as an example of how `immer` could more fully replace our use of `immutable`.

First, is to keep the `[immerable] = true;` declaration on the class. As stated in the [immer docs on working with classes](https://immerjs.github.io/immer/docs/complex-objects) this is how a JS class can be annotated so it is "draftable" by immer. This allows the class to be mutated properly within immer's `produce` method. If this isn't done then the class instance will be mutated directly (rendering call to `produce` useless -- as expected).

Second, is to make `QueryModel` have all `readonly` fields. This allows TypeScript to enforce at compile time and JS to enforce at runtime (say if someone works around it with @ts-ignore, casts to all, etc). With this no one can modify an instance and cause unintended side-effects.

```ts
export class QueryModel implements IQueryModel {
    [immerable] = true;

    // Fields from QueryConfig
    readonly baseFilters: Filter.IFilter[];
    readonly containerFilter?: Query.ContainerFilter;
    readonly containerPath?: string;
    readonly id: string;
    readonly includeDetailsColumn: boolean;
    readonly includeUpdateColumn: boolean;
    readonly keyValue?: any; // TODO: better name
    readonly maxRows?: number;
    readonly offset: number;
    readonly omittedColumns: string[];
    readonly queryParameters?: { [key: string]: any};
    readonly requiredColumns: string[];
    readonly schemaQuery: SchemaQuery;
    readonly sorts?: QuerySort[];

    // QueryModel only fields
    readonly error: string;
    readonly filterArray: Filter.IFilter[];
    readonly rowsLoadingState: LoadingState;
    readonly messages: GridMessage[];
    readonly orderedRows?: string[];
    readonly queryInfo?: QueryInfo;
    readonly queryInfoLoadingState: LoadingState;
    readonly rows?: { [key: string]: any};
    readonly rowCount?: number;

    ...
```

Third, is that any `produce` calls that expect to mutate a `QueryModel` will need to specify the type as `Draft<Type>` as specified in the [immer docs on using with TypeScript](https://immerjs.github.io/immer/docs/typescript):
> The `Draft` utility type can be used if the state argument type is immutable.

Example of this change done in `withQueryModels.tsx`:

```ts
this.setState(produce((draft: Draft<State>) => {
    draft.queryModels[id].queryInfoLoadingState = LoadingState.LOADING;
}));
```

Without this TypeScript will still not allow writing to a readonly property.

With these three changes I was able to get this fully working locally against your changes. I realize this makes it harder to test the model directly and ended up choosing to add a `mutate` method to the model directly making it easier for tests to make changes. This method could also be exposed only in tests as a helper if you prefer not to expose it on the model class itself.

```ts
/**
 * Returns a deep copy of this model with props applied iff props is not empty/null/undefined else
 * returns this.
 * @param props
 */
mutate(props: Partial<IQueryModel>): QueryModel {
    return produce(this, (draft: Draft<QueryModel>) => {
        Object.assign(draft, props);
    });
}
```

I've DM'ed you a patch containing these changes plus fixes for the tests.

Reply:
Patch applied.